src/writer.jl

```
@@ -0,0 +1,528 @@
+    uncompressed_byte_data = reinterpret(UInt8, colvals) |> collect
+
+    if codec == PAR2.CompressionCodec.UNCOMPRESSED
+        return uncompressed_byte_data
+    elseif codec == PAR2.CompressionCodec.SNAPPY
+        compressed_data = Snappy.compress(uncompressed_byte_data)
+    elseif codec == PAR2.CompressionCodec.GZIP
+        compressed_data = transcode(GzipCompressor, uncompressed_byte_data)
+    elseif codec == PAR2.CompressionCodec.LZ4
+        error("lz4 is not supported as data compressed with https://github.com/JuliaIO/CodecLz4.jl can't seem to be read by R or Python. If you know how to fix it please help out.")
+        #compressed_data = transcode(LZ4HCCompressor, uncompressed_byte_data)
+    elseif codec == PAR2.CompressionCodec.ZSTD
+        compressed_data = transcode(ZstdCompressor, uncompressed_byte_data)
+    else
+        error("not yet implemented")
+    end
+
+    return compressed_data
+end
+
+function compress_using_codec(colvals::AbstractVector{String}, codec::Int)::Vector{UInt8}
+    """Compress `String` column using codec"""
+    # the output
+    io = IOBuffer()
+
+    # write the values
+    for val in colvals
+        # for string it needs to be stored as BYTE_ARRAY which needs the length
+        # to be the first 4 bytes UInt32
+        write(io, val |> sizeof |> UInt32)
+        # write each of the strings one after another
+        write(io, val)
+    end
+
+    uncompressed_bytes = take!(io)
+    return compress_using_codec(uncompressed_bytes, codec)
+end
+
+function write_col_dict(fileio, colvals::AbstractArray{T}, codec) where T
+    """ write the column dictionary page """
+    # note: `level`s does not return `missing` as a level
+    uvals = DataAPI.levels(colvals)
+
+    # do not support dictionary with more than 127 levels
+    # TODO relax this 127 restriction
+    if length(uvals) > 127
+        @warn "More than 127 levels in dictionary. This is not supported at this stage."
```

should probably throw an error?

---

I think we shouldn't throw error here as it's just something no supported by Parquet.jl at this stage. It's actually supported by Parquet.

---

Warning may have been fine if the data were actually written, but suboptimally. For example, warning that there's no CRC being calculated now is probably fine, because CRC may be considered as an optional check.

But since here it skips writing some data altogether, there is a risk of the user trying to use the resulting incomplete file further down in their workflow. It is better to throw an error IMO.

---

It doesn't write a corrupted file, if that's your concern. It is actually fine to skip the writing of the dictionary. In this first version, dictionary pages are not used at all. In the next version, we will use dictionary pages properly.

---

@tanmaykm Are you OK to merge it as it for this version? If not I can remove the dictionary writing code altogether this round. And include it in next version without that 127 level restriction.

As I said above, currently the dictionary is not used in writing anyway, so it's fine as it is. But throwing an error is definitely not correct as the file doesn't need a dictionary.

---

Ah okay, so `write_col_dict` is not called anywhere? I always find it less confusing to comment out unused code.

---

commented out in latest push (soon)